### PR TITLE
CRD changes for monitoring

### PR DESCRIPTION
**Pull request by khansaad**
Signed-off-by: Saad Khan <saakhan@redhat.com>

This PR contains following changes:

- Updated dockerFile to remove dependency on the rolebindings, service account etc when deploying on openshift.
- Moved the configs and CRD files in a new autotune directory to keep it separate from monitoring yaml.
- Handled watcher exception when no CRDs are applied in case of monitoring.
- Added a new file  `kruize-monitoring-openshift.yaml` to include all the configs and CRDs as part of single file for monitoring use case ( Still in progress.. )

### CODE REVIEW

**Comment by khansaad** [comment]
**App_logs :** 
[kruize-openshift-monitoring-app-logs.log](https://github.com/kruize/autotune/files/10192516/kruize-openshift-monitoring-app-logs.log)

**After running the monitoring demo -**

> Result logs on openshift :

[Openshift_Demo_Test_logs.log](https://github.com/kruize/autotune/files/10192520/Openshift_Demo_Test_logs.log)

> Result logs on Minikube:

[Minikube_Demo_Test_logs.log](https://github.com/kruize/autotune/files/10192523/Minikube_Demo_Test_logs.log)

File: deploy.sh [hunk 1]
```diff
@@ -18,23 +18,24 @@
...
+AUTOTUNE_QUERY_VARIABLE_CRD="manifests/autotune/autotune-query-variable-crd.yaml"
+AUTOTUNE_DEPLOY_MANIFEST_TEMPLATE="manifests/autotune/autotune-operator-deployment.yaml_template"
+AUTOTUNE_DEPLOY_OPENSHIFT_MANIFEST_TEMPLATE="manifests/autotune/autotune-operator-openshift-deployment.yaml_template"
+AUTOTUNE_DEPLOY_MANIFEST="manifests/autotune/autotune-operator-deployment.yaml"
+KRUIZE_DEPLOY_MANIFEST="manifests/monitoring/kruize-monitoring-openshift.yaml"
+AUTOTUNE_RB_MANIFEST_TEMPLATE="manifests/autotune/autotune-operator-rolebinding.yaml_template"
+AUTOTUNE_RB_MANIFEST="manifests/autotune/autotune-operator-rolebinding.yaml"
+AUTOTUNE_ROLE_MANIFEST="manifests/autotune/autotune-operator-role.yaml"
+AUTOTUNE_SA_MANIFEST="manifests/autotune/autotune-operator-sa.yaml"
+SERVICE_MONITOR_MANIFEST="manifests/autotune/servicemonitor/autotune-service-monitor.yaml"
```
Comment (dinogun) [inline deploy.sh:32]: Add a var for `manifests/autotune` and `manifests/monitoring` and reference them for all the variables here and elsewhere in the script

File: manifests/monitoring/kruize-monitoring-minikube.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  replicas: 1
+  selector:
+      matchLabels:
+        name: autotune
+  template:
+    metadata:
+      labels:
+        app: autotune
+        name: autotune
+        operatorframework.io/os.linux: supported
```
Comment (dinogun) [inline manifests/monitoring/kruize-monitoring-minikube.yaml:18]: Why is this needed?

File: manifests/monitoring/kruize-monitoring-openshift.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          - name: MONITORING_AGENT
+            value: "prometheus"
+          - name: MONITORING_SERVICE
+            value: "prometheus-k8s"
+          - name: MONITORING_AGENT_ENDPOINT
+            value: "prometheus-k8s"
+          - name: LOGGING_LEVEL
+            value: "info"
+          - name: ROOT_LOGGING_LEVEL
+            value: "error"
```
Comment (dinogun) [inline manifests/monitoring/kruize-monitoring-openshift.yaml:42]: Should this be error or info?

File: manifests/monitoring/kruize-monitoring-openshift.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: autotune
+  labels:
+    app: autotune
+  namespace: openshift-tuning
```
Comment (dinogun) [inline manifests/monitoring/kruize-monitoring-openshift.yaml:7]: Looks like this is assuming that the namespace exists. Need to add the namespace creation as well

File: .github/workflows/test-on-pr.yaml [hunk 5]
```diff
@@ -43,10 +43,10 @@ jobs:
       - name: Deploy autotune
         run: |
           echo Deploy autotune
-          cp ./manifests/autotune-operator-deployment.yaml_template ./manifests/autotune-operator-deployment.yaml_template.old
-          sed -e "s/imagePullPolicy: Always/imagePullPolicy: IfNotPresent/g" ./manifests/autotune-operator-deployment.yaml_template.old > ./manifests/autotune-operator-deployment.yaml_template
+          cp ./manifests/autotune/autotune-operator-deployment.yaml_template ./manifests/autotune/autotune-operator-deployment.yaml_template.old
+          sed -e "s/imagePullPolicy: Always/imagePullPolicy: IfNotPresent/g" ./manifests/autotune/autotune-operator-deployment.yaml_template.old > ./manifests/autotune/autotune-operator-deployment.yaml_template
```
Comment (dinogun) [inline .github/workflows/test-on-pr.yaml:47]: How are we testing the monitoring mode?

File: src/main/java/com/autotune/common/target/kubernetes/service/impl/KubernetesServicesImpl.java [hunk 6]
```diff
@@ -420,7 +420,7 @@ public Event getEvent(String namespace, String eventName) {
         try {
             event = kubernetesClient.events().inNamespace(namespace).withName(eventName).get();
         } catch (Exception e) {
-            new TargetHandlerException(e, "getEvent failed!");
+            LOGGER.error("Watcher not added! Only Monitoring APIs are exposed.");
```
Comment (dinogun) [inline src/main/java/com/autotune/common/target/kubernetes/service/impl/KubernetesServicesImpl.java:423]: This should be a `LOGGER.warn`. The message should read, `Watcher not added! Only REST API access is enabled"

Comment (khansaad) [reply src/main/java/com/autotune/common/target/kubernetes/service/impl/KubernetesServicesImpl.java]: Done

File: manifests/monitoring/kruize-monitoring-openshift.yaml (hunk 4 above)
Comment (khansaad) [reply manifests/monitoring/kruize-monitoring-openshift.yaml:12]: Added

File: manifests/monitoring/kruize-monitoring-minikube.yaml (hunk 2 above)
Comment (khansaad) [reply manifests/monitoring/kruize-monitoring-minikube.yaml]: Removed

File: manifests/monitoring/kruize-monitoring-minikube.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: autotune
```
[the remark this answers is not in the document]
Comment (khansaad) [reply manifests/monitoring/kruize-monitoring-minikube.yaml]: Done

File: deploy.sh (hunk 1 above)
Comment (khansaad) [reply deploy.sh]: Done

File: .github/workflows/test-on-pr.yaml (hunk 5 above)
Comment (khansaad) [reply .github/workflows/test-on-pr.yaml:47]: Updated the yaml now to include monitoring as well.

File: deploy.sh [hunk 8]
```diff
@@ -18,23 +18,28 @@
...
-AUTOTUNE_DEPLOY_MANIFEST_TEMPLATE="manifests/autotune-operator-deployment.yaml_template"
-AUTOTUNE_DEPLOY_OPENSHIFT_MANIFEST_TEMPLATE="manifests/autotune-operator-openshift-deployment.yaml_template"
-AUTOTUNE_DEPLOY_MANIFEST="manifests/autotune-operator-deployment.yaml"
-AUTOTUNE_RB_MANIFEST_TEMPLATE="manifests/autotune-operator-rolebinding.yaml_template"
-AUTOTUNE_RB_MANIFEST="manifests/autotune-operator-rolebinding.yaml"
-AUTOTUNE_ROLE_MANIFEST="manifests/autotune-operator-role.yaml"
-AUTOTUNE_SA_MANIFEST="manifests/autotune-operator-sa.yaml"
-SERVICE_MONITOR_MANIFEST="manifests/servicemonitor/autotune-service-monitor.yaml"
+AUTOTUNE_DIR="./manifests/autotune"
+MONITORING_DIR="./manifests/monitoring"
```
Comment (dinogun) [inline deploy.sh:22]: As I mentioned yesterday this should be `crc` not monitoring

Comment (dinogun) [reply deploy.sh:22]: Also change MONITORING_DIR to `CRC_DIR`

File: manifests/monitoring/kruize-monitoring-openshift.yaml (hunk 3 above)
Comment (dinogun) [reply manifests/monitoring/kruize-monitoring-openshift.yaml]: Initially we want INFO messages to be turned on a well as this will aid in debugging

**Comment by dinogun** [comment]
@khansaad Please rebase to latest mvp_demo as have added the profiles dir in manifests now

File: deploy.sh (hunk 8 above)
Comment (khansaad) [reply deploy.sh]: Done

File: manifests/monitoring/kruize-monitoring-openshift.yaml (hunk 3 above)
Comment (khansaad) [reply manifests/monitoring/kruize-monitoring-openshift.yaml]: Changed to INFO

**Comment by dinogun** [comment]
The `performance-profiles` dir also needs to be inside the `autotune` dir

**Review by dinogun — APPROVED** [review]
LGTM

**Comment by dinogun** [comment]
@khansaad It looks like the rebase was not done right :( I see that we now have two copies of the performance profile yamls

**Comment by dinogun** [comment]
#541 added perf profiles and it looks like it was duplicated here.
